astropy/utils/parsing.py

```diff
@@ -27,7 +37,7 @@
 _LOCK = threading.RLock()
 
 
-def _add_tab_header(filename, package):
+def _add_tab_header(filename: str, package: str) -> None:
```

Comment:
These annotations are correct, but I not that internally, the function binds the name `f` twice *to objects with different types* (file buffers with different permissions), which mypy will flag as soon as we turn it on, so I suggest fixing it now rather than later (it suffices to use 2 different variable names).

Reply:
I'd prefer to keep type annotation pull requests free from any code changes. If the code should be updated then it's better to do that in a separate pull request. Besides, changing the name of one variable does not sound like a big improvement to me. I would rather rewrite this function in its entirety:
```python
def _add_tab_header(filename: Path, package: str) -> None:
    contents = filename.read_text()
    filename.write_text(_TAB_HEADER.format(package=package))
    filename.open('a').write(contents)
```

Reply:
Agreed with @eerovaher here; better to have no code changes in typing PRs (and, really, if `mypy` cannot handle temporary variables like here, then perhaps it needs to be fixed...)

Reply:
I don't think any static analyzer handles mutating types without a prior declaration of a type union.

e..g

```python
a: Union[str, int]
a = "a"
a = 1
```

I also don't think that's something that should be fixed!
Dynamism, like most tools, is both useful and dangerous. I think mutating types in-place is more on the dangerous side than useful. Renaming a variable is both easy and safe. Static type checks should catch this.

Reply:
Agreed with @nstarman . I mostly wanted to raise awareness that by adding annotations *before* enabling type-checking would create technical debt in the form of *more* errors to be fixed to get type-checking started instead of less (well, not sure what the net change will be). But the current approach is still better than not addressing type hints at all, and it reached some kind of consensus.

Reply:
I also note, for closure, that @eerovaher's effort in reducing long functions might compensate for this specific kind of new "errors" at type-checking time: the probability that a variable name be used for more than one type is greatly increased in long scopes.